docs/layout.rst

```diff
@@ -0,0 +1,155 @@
+.. _layout:
+
+Tensor layout
+=============
+
+Tensors (:ref:`scalar_tensor`) can be *placed* in a specific shape and *layout*.
+Having a good layout can be the key to performance.
+
+In Taichi, placing a layout is treated in a recursive manner.
+
+For example, this declares a 0-D tensor:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32)
+    ti.root.place(x)
+    # or
+    x = ti.var(ti.f32, shape=())
+
+This declares a 1-D tensor of size ``3``:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32)
+    ti.root.dense(ti.i, 3).place(x)
+    # or
+    x = ti.var(ti.f32, shape=3)
+
+This declares a 1-D tensor of shape ``(3, 4)``:
+
+.. code-block:: python
+
+    x = ti.var(ti.f32)
+    ti.root.dense(ti.ij, (3, 4)).place(x)
+    # or
+    x = ti.var(ti.f32, shape=(3, 4))
+
+Now, you may say, why not just simply specify ``shape=`` argument to specify tensor shape? Why we need these fancy stuffs like ``ti.root`` and ``place``? Wasn't them the same?
+Good question, let go forward and figure out why.
+
+
+Multi-dimentional layout
+------------------------
+
+As we all know, most mordern computers have a 1-D memory model.
```

Comment:
"""
Let's start with the simplest layout.

Since address space are linear in most modern architectures, for 1D Taichi tensors, the address of ``i``-ith element is simply ``i``.
"""